src/limits/MemoryLimitListener.cc

```diff
@@ -40,33 +40,52 @@ MemoryLimitListener::MemoryLimitListener(uint64_t memoryLimitKb)
                     if (!vmPeakValid_) {
                         return tracer::TraceAction::CONTINUE;
                     }
-
-                    uint64_t memoryUsage = getMemoryUsageKb() +
-                                           tracee.getSyscallArgument(1) / 1024;
-                    memoryPeakKb_ = std::max(memoryPeakKb_, memoryUsage);
-                    outputBuilder_->setMemoryPeak(memoryPeakKb_);
-                    logger::debug(
-                            "Memory usage after mmap ",
-                            VAR(memoryUsage),
-                            ", ",
-                            VAR(memoryPeakKb_));
-
-                    if (memoryUsage > memoryLimitKb_) {
-                        outputBuilder_->setKillReason(
-                                printer::OutputBuilder::KillReason::MLE,
-                                "memory limit exceeded");
-                        logger::debug(
-                                "Limit ",
-                                VAR(memoryLimitKb_),
-                                " exceeded, killing tracee");
-                        return tracer::TraceAction::KILL;
-                    }
-                    return tracer::TraceAction::CONTINUE;
+                    return handleMemoryAllocation(tracee.getSyscallArgument(1) / 1024);
                 }),
                 Arg(0) == 0 && Arg(1) > MEMORY_LIMIT_MARGIN / 2));
     }
+    syscallRules_.emplace_back(seccomp::SeccompRule(
+            "mremap",
+            seccomp::action::ActionTrace([this](tracer::Tracee& tracee){
+                TRACE();
+                if (!vmPeakValid_) {
+                    return tracer::TraceAction::CONTINUE;
+                }
+                bool doesUnmap = (tracee.getSyscallArgument(3)&MREMAP_DONTUNMAP)!=0;
+                // Allow user to shrink its memory
+                if(!doesUnmap&&tracee.getSyscallArgument(1)>=tracee.getSyscallArgument(2)){
+                    return tracer::TraceAction::CONTINUE;
+                }
+                uint64_t newMemoryAllocated=tracee.getSyscallArgument(2);
+                // Do not count already allocated memory
+                if(!doesUnmap) newMemoryAllocated-=tracee.getSyscallArgument(1);
```

Comment:
This syscall has many arguments, and you're using them multiple times, so it gets a bit confusing to read.

Maybe we could have variables at the top of this function giving names to these arguments, like
```
uint64_t old_size = tracee.getSyscallArgument(1);
uint64_t new_size = tracee.getSyscallArgument(2);
uint64_t flags = tracee.getSyscallArgument(3);
```
and use them throughout the function, to make it more readable?